### PR TITLE
Update the introduction of step-17.

### PR DESCRIPTION
**Pull request by bangerth**
It's late once again, but I wanted to put this up to get some feedback.
While looking into #1365, I realized that step-17's introduction is not
particularly helpful. This is the first MPI program, and some more
background would be helpful. So this is what it provides.

I also update a glossary entry that I though was not particularly
well written. It probably still isn't, but the phrase
'associated with' is, in my mind, not as clear as 'owned by'.
Feedback on this welcome as well.

### CODE REVIEW

File: doc/doxygen/headers/glossary.h [hunk 1]
```diff
@@ -1299,25 +1299,26 @@
...
  * communicator). Partitioning is typically done using the
  * GridTools::partition() function, but any other method can also be used to
- * do this.
+ * do this. (Alternatively, the parallel::shared::Triangulation class can
+ * partition the mesh automatically using a similar approach.)
  *
  * On the other hand, for programs that are parallelized using MPI but
  * where meshes are held distributed across several processors using
  * the parallel::distributed::Triangulation and
  * parallel::distributed::DoFHandler classes, the subdomain id of
```
Comment (kronbichler) [inline doc/doxygen/headers/glossary.h]: This is not actually part of the commit, but we have retired parallel::distributed::DoFHandler long ago. It was even before we merged the parallel distributed branch IIRC. Just remove that part.

File: examples/step-17/doc/intro.dox [hunk 2]
```diff
@@ -13,58 +14,169 @@ within an MPI network, the resulting code will even be able to solve the
...
-near-equivalents <code>PETScWrappers::Vector</code> and
-<code>PETScWrappers::SparseMatrix</code> (for sequential vectors and matrices,
-i.e. objects for which all elements are stored locally on one machine), and
-<code>PETScWrappers::MPI::Vector</code> and
-<code>PETScWrappers::MPI::SparseMatrix</code> for versions of these classes
-where only a part of the matrix or vector is stored on each machine within an
-MPI network. These classes provide an interface that is very similar to that
+near-equivalents <code>PETScWrappers::MPI::Vector</code> and
+<code>PETScWrappers::MPI::SparseMatrix</code> that store data in a way so that
+every processor in the MPI network on stores
```
Comment (kronbichler) [inline examples/step-17/doc/intro.dox]: on -> only

File: examples/step-17/doc/intro.dox [hunk 3]
```diff
@@ -13,58 +14,169 @@ within an MPI network, the resulting code will even be able to solve the
...
+a part of the matrix or vector. More specifically, each processor will
+only store those rows of the matrix that corresponds to a degree of
```
Comment (kronbichler) [inline examples/step-17/doc/intro.dox]: corresponds -> correspond

File: examples/step-17/doc/intro.dox [hunk 4]
```diff
@@ -13,58 +14,169 @@ within an MPI network, the resulting code will even be able to solve the
...
+that multiple processors can all access all data within one memory
+space, for example whenever multiple cores in a single machine work
+on a common task). Some of the fundamentals of distributed memory
+computing are discussed in the 
+@ref distributed "Parallel computing with multiple processors using distributed memory"
+documentation module, which is itself a sub-module of the
+@ref Parallel "Parallel computing" module. 
+
+In general, to be truly able to scale to large numbers of processors, one
+needs to split <i>every</i> data structure within a program between the
```
Comment (kronbichler) [inline examples/step-17/doc/intro.dox]: I would suggest to slightly reword this. We split every *global* data structure that is proportional to the size of the problem, but replicate local things such as FiniteElement, the local matrices and vectors, and so on. (One could imagine a case where those data structures are also split but it requires yet another mindset.)

File: examples/step-17/doc/intro.dox [hunk 5]
```diff
@@ -13,58 +14,169 @@ within an MPI network, the resulting code will even be able to solve the
...
+cluster. But under the hood, whenever those processors become available,
+the same call as above will generally be executed.) What this does is that
+the MPI system will start 32 <i>copies</i> of the <code>step-17</code>
+executable. This may happen on different machines that can't even read
+from each others' memory spaces, or it may happen on the same machine, but
+the end result is the same: each of these 32 copies will run with some
+memory allocated to it by the operating system, and it will not directly
+be able to read the memory of the other 31 copies. In order to collaborate
+in a common task, these 32 copies then have to <i>communicate</i> with
+each other. MPI, short for <i>Message Passing Interface</i> makes this
```
Comment (kronbichler) [inline examples/step-17/doc/intro.dox]: insert a comma before "makes"

File: examples/step-17/doc/intro.dox [hunk 6]
```diff
@@ -13,58 +14,169 @@ within an MPI network, the resulting code will even be able to solve the
...
+possible by allowing programs to <i>send messages</i>. You can think
+of this as the mail service: you can put a letter to a specific address
+into the mail and it will be delivered. But that's the extent to which
+you can control things. If you want the received to do something
```
Comment (kronbichler) [inline examples/step-17/doc/intro.dox]: received -> receiver

**Comment by kronbichler** [comment]
This looks good. Please fix the remaining typos and merge yourself.

File: doc/doxygen/headers/glossary.h (hunk 1 above)
Comment (bangerth) [inline doc/doxygen/headers/glossary.h]: I moved this to #2018.

File: examples/step-17/doc/intro.dox (hunk 4 above)
Comment (bangerth) [inline examples/step-17/doc/intro.dox]: Excellent point.

**Pull request merged**